### PR TITLE
Enable building using LLVM_EXTERNAL_PROJECTS.

### PR DESCRIPTION
**Pull request by mikeurbach**
This allows building NPCOMP as an external project of LLVM, similar to
how CIRCT can be built: https://github.com/llvm/circt/pull/227.

The CMake options to use this build style look like this:

```
  -DLLVM_EXTERNAL_PROJECTS=npcomp \
  -DLLVM_EXTERNAL_NPCOMP_SOURCE_DIR=/path/to/mlir-npcomp \
```

### CODE REVIEW

File: CMakeLists.txt
Comment (mikeurbach) [inline CMakeLists.txt:196]: The `check-all` target seemed to conflict with the MLIR one. I added this `npcomp` namespacing, but I'm not sure if there is a better way.

File: frontends/pytorch/test/lit.cfg.py
Comment (mikeurbach) [inline frontends/pytorch/test/lit.cfg.py:61]: I found that the `npcomp` python artifacts were placed under `tools/npcomp/python` when doing an external build. A couple things about this: 1) I'm not sure if there is a better way to use CMake to put the python just in `build/python` when doing an external build, 2) I think the `npcomp` in the `tools/npcomp/python` path comes from the name passed to the `LLVM_EXTERNAL_PROJECTS` flag, so this may not be guaranteed to match for everyone's setup. I have a feeling there is a better way to achieve this line of python with CMake, but I haven't figured it out yet.

File: frontends/pytorch/test/lit.site.cfg.py.in
Comment (mikeurbach) [inline frontends/pytorch/test/lit.site.cfg.py.in:54]: I think this change is a safe way to make this path agnostic to CMAKE_SOURCE_DIR, which can be different for an external build. It works on my computer and the GitHub check, but please correct me if I'm wrong.

File: include/npcomp/Conversion/CMakeLists.txt
Comment (mikeurbach) [inline include/npcomp/Conversion/CMakeLists.txt:5]: This was conflicting with the MLIR ConversionPasses name.

File: frontends/pytorch/test/lit.cfg.py
Comment (stellaraccident) [inline frontends/pytorch/test/lit.cfg.py:61]: Side comment (nothing to do about it here): I find this layering of such projects under 'tools' to be really annoying but haven't taken the time to see what would need to be done to pop it up a level.

**Review by stellaraccident — APPROVED** [review]
Thanks - been on my list to figure this out for a long time, and it was comparatively more changes than I expected. Thanks again.

**Review by mikeurbach — COMMENTED** [review]
I have added this in a similar fashion to the linked CIRCT PR, and called out some of the NPCOMP-specific tweaks to the general recipe. I've been building both projects for about a month in this manner, which I do enjoy (one command to build LLVM/MLIR/CIRCT/NPCOMP). I would like to upstream the tweaks necessary to mlir-npcomp if possible. It is by no means urgent, but I wanted to take the time to make a PR instead of carrying this locally.

I am really not that familiar with CMake, and I basically just followed what was done in CIRCT. We don't do a full build and test in GitHub for the external build, but we do have a step that runs the configure command to make sure it doesn't barf. I could add something similar if that would be of interest here.

This PR may benefit from hiding whitespace changes during review: https://github.com/llvm/mlir-npcomp/pull/152/files?w=1

File: CMakeLists.txt
Comment (stellaraccident) [reply CMakeLists.txt:196]: sgtm

File: frontends/pytorch/test/lit.cfg.py
Comment (stellaraccident) [reply frontends/pytorch/test/lit.cfg.py:61]: Let's land it in a working state how you have it. Then I/someone can try to find a more targeted place to put it. I'd hope we can just emit all of the python artifacts into the top level `build/python` directory: even across projects, they form a namespace that must not overlap, so having everything write to that directory is ok (CMake will complain if two projects then try to write the same thing into the python namespace).

File: frontends/pytorch/test/lit.site.cfg.py.in
Comment (stellaraccident) [reply frontends/pytorch/test/lit.site.cfg.py.in:54]: I'm ok with it if it works. Can't think of anything better without fiddling with it.

File: include/npcomp/Conversion/CMakeLists.txt
Comment (stellaraccident) [reply include/npcomp/Conversion/CMakeLists.txt:5]: Thx. Copy-pasta from the beginning of the project.

**Comment by mikeurbach** [comment]
Wow @stellaraccident you reviewed faster than I could request it. I do wonder if there may be some CMake incantation to handle the layering under 'tools' differently. If you are OK with this as-is, I will leave that to future spelunking.

**Comment by stellaraccident** [comment]
> Wow @stellaraccident you reviewed faster than I could request it. I do wonder if there may be some CMake incantation to handle the layering under 'tools' differently. If you are OK with this as-is, I will leave that to future spelunking.

Caught me at an opportune time :)

If it works, let's commit as-is. Incremental CMake work is the name of the game... we can shift things later with smaller changes once we figure out how to do them.

**Comment by mikeurbach** [comment]
Sounds good to me. Other than the GitHub checks and using external builds locally, is there anything else I can test before I land these changes?

**Comment by stellaraccident** [comment]
Not that I can think of, but if it is not too inconvenient, could you land
(or have one of us land) at a time during US working hours and post a note
to #mlir-npcomp to have one of us try it? I'm happy to hit the merge button
if not convenient for you.

**Comment by mikeurbach** [comment]
Yep I can do that during the day tomorrow. I believe I have permission here to hit merge.

**Pull request merged**
